internal/lang/eval.go

```diff
@@ -489,3 +492,87 @@ func normalizeRefValue(val cty.Value, diags tfdiags.Diagnostics) (cty.Value, tfd
+// the user on how to remedy the situation in Terraform, and we can give more
+// useful information in a few Terraform specific cases here.
+func checkForUnknownFunctionDiags(diags hcl.Diagnostics) hcl.Diagnostics {
+	for _, d := range diags {
+		extra, ok := hcl.DiagnosticExtra[hclsyntax.FunctionCallUnknownDiagExtra](d)
+		if !ok {
+			continue
+		}
+		name := extra.CalledFunctionName()
+		namespace := extra.CalledFunctionNamespace()
+		namespaceParts := strings.Split(namespace, "::")
+		if len(namespaceParts) < 2 {
+			// no namespace (namespace includes ::, so will have at least 2
+			// parts), but check if there is a matching name in a provider
+			// namspace.
+			if d.EvalContext == nil {
+				continue
+			}
+
+			for funcName := range d.EvalContext.Functions {
+				if strings.HasSuffix(funcName, "::"+name) {
+					d.Detail = fmt.Sprintf("%s Did you mean %q?", d.Detail, funcName)
+					break
+				}
+			}
+			continue
+		}
+
+		// the diagnostic isn't really shared with anything, and copying would
+		// still retain the internal pointers, so we're going to modify the
+		// diagnostic in-place if we want to change the output. Log the original
+		// diagnostic for debugging purposes in case we overwrite something
+		// potentially useful in the future from hcl.
+		log.Printf("[ERROR] UnknownFunctionCall: %s", d.Error())
+		d.Summary = "Unknown provider function"
+
+		if namespaceParts[0] != "provider" {
+			// help if the user is skipping the provider:: prefix before the
+			// provider name.
+			d.Detail = fmt.Sprintf(`The function namespace %q is not valid. Provider function calls must use the "provider::" namespace prefix.`, namespaceParts[0])
+			continue
+		}
+
+		if namespaceParts[1] == "" {
+			// missing provider name entirely
+			d.Detail = `The function call must include the provider name after the "provider::" prefix.`
+			continue
+		}
+
+		if d.EvalContext == nil {
+			// There's no eval context for some reason, so we can't inspect the
+			// available functions.
+			d.Detail = fmt.Sprintf(`There is no function named "%s%s".`, namespace, name)
+			continue
+		}
+
+		otherProviderFuncs := false
+		for funcName := range d.EvalContext.Functions {
+			// there are other functions in this provider namespace, so it must
+			// have been included in the configuration, and we can be clear that
+			// this a function which the provider does not support.
+			if strings.HasPrefix(funcName, namespace) {
+				otherProviderFuncs = true
+				break
+			}
+		}
+		if otherProviderFuncs {
+			d.Detail = fmt.Sprintf("The function %q is not available from the provider %q.", name, namespaceParts[1])
+			continue
+		}
+
+		// no other functions exist for this provider, so hint that the user may
+		// need to include it in the configuration.
+		d.Detail = fmt.Sprintf(`There is no function named "%s%s". The provider %q may need to be added to the required_providers block within the module configuration.`, namespace, name, namespaceParts[1])
```

Comment:
I suggest active-voice and more assertive wording here, to make it clearer that this is an action the user must take, not an action that Terraform might take, or similar.

"You must declare provider local name %q in this module's required_providers block."

I sense from the comment above that this message would also appear for an attempt to call a function in a provider that offers no functions, and so this message was trying to hedge a little. While of course I would prefer if we could find a way to distinguish those two cases with separate error messages, I can see why that would be difficult here in which case I'd propose the following compromise:

"Ensure that provider name %q is declared in this module's required_providers block, and that this provider offers a function named %q."

That way it covers both of the possibilities while still erring on the side of directly telling the reader what they need to do, to ensure the message is actionable.

Reply:
Thank you! That message sounds much better